src/core/splitter/EdgeGraph.js

```diff
@@ -0,0 +1,688 @@
+					insertedEdge.end.copy( triangle.points[ nextEdgeIndex ] );
+					insertedEdge.endIndex = triangle.getVertexIndex( nextEdgeIndex );
+
+					const finalEdgeIndex0 = ( edgeIndex + 1 ) % 3;
+					const newTri0 = trianglePool.getInstance();
+					newTri0.setEdge( 0, l0, true );
+					newTri0.setEdge( 1, triangle.edges[ finalEdgeIndex0 ].edge, triangle.edges[ finalEdgeIndex0 ].reversed );
+					newTri0.setEdge( 2, insertedEdge, true );
+
+					const finalEdgeIndex1 = ( edgeIndex + 2 ) % 3;
+					const newTri1 = trianglePool.getInstance();
+					newTri1.setEdge( 0, l1, true );
+					newTri1.setEdge( 1, insertedEdge, false );
+					newTri1.setEdge( 2, triangle.edges[ finalEdgeIndex1 ].edge, triangle.edges[ finalEdgeIndex1 ].reversed );
+
+					triangles.splice( triangles.indexOf( triangle ), 1 );
+					triangles.push( newTri0, newTri1 );
+					edges.push( insertedEdge );
+
+				}
+
+				edges.push( l0, l1 );
+				edges.splice( intersectingEdge, 1 );
+
+			}
+
+		}
+
+		return index;
+
+	}
+
+	findMatchingPointIndex( p ) {
+
+		// find the matching vertex for the give point if it exists in the graph
+		const points = this.points;
+		let closestIndex = null;
+		let closestDist = Infinity;
+		for ( let i = 0, l = points.length; i < l; i ++ ) {
+
+			const d = p.distanceTo( points[ i ] );
+			if ( d < EPSILON && d < closestDist ) {
+
+				closestIndex = i;
+				closestDist = d;
+
+			}
+
+		}
+
+		return closestIndex;
+
+	}
+
+	markMatchingEdgeRequired( inserting ) {
+
+		const { edges } = this;
+
+		_edgesToSwap.length = 0;
+		for ( let i = 0, l = edges.length; i < l; i ++ ) {
+
+			// swap the edge if we don't emanate from the same point
+			const other = edges[ i ];
+			if (
+				other.startIndex !== inserting.startIndex &&
+				other.startIndex !== inserting.endIndex &&
+				other.endIndex !== inserting.startIndex &&
+				other.endIndex !== inserting.endIndex
+			) {
+
+				// TODO
+				// if ( areEdgesParallel( inserting, other ) ) {
+
+				// 	let sp = inserting.closestPointToPointParameter( other.start, false );
+				// 	let ep = inserting.closestPointToPointParameter( other.end, false );
+
+				// 	if ( ! (
+				// 		sp < 0 && ep < 0 ||
+				// 		sp > 1 && ep > 1
+				// 	) ) {
+
+				// 		// TODO: but do they overlap
+				// 		// console.log( 'THEYRE PARALLEL', sp, ep );
+
+				// 	}
+
+				// }
+
+				// TODO: make sure we account for parallel scenarios
+				if ( lineIntersect( inserting, other, _vec ) ) {
+
+					_edgesToSwap.push( other );
+
+				}
+
+			} else {
+
+				// if we found an edge that matches without swapping then theres no need
+				// to continue
+				if ( doEdgesMatch( inserting, other ) ) {
+
+					other.required = true;
+					return true;
+
+				}
+
+			}
+
+		}
+
+		// try for a few iterations to swap edges until they work
+		for ( let i = 0; i < SWAP_ITERATIONS; i ++ ) {
+
+			for ( let j = 0, l = _edgesToSwap.length; j < l; j ++ ) {
+
+				const other = _edgesToSwap[ j ];
+				this.swapEdge( other );
```

Comment:
Another issue we ran into. 

Swapping this edge can lead to one of the triangles to be a 0 area triangle, which then makes it not able to clean up.

I'm not sure if there is a downstream place to fix it, but we discovered this would lead to some invalid geometry.

If I check if the resulting triangles are not 0 area this ends up not being a problem

Reply:
Actually I looked into this more and the problem is a bit more generic.

This function swaps edges, which can lead to an invalid triangle or a flipped triangle.

So we need to detect if the swapped edge will interect or flip over the none change edge for both of the attached triangles:

I can improve this by adding this to swapEdge
```
const otherEdgeIndex = ( t0SwapIndex + 1 ) % 3;
		const otherEdgeIndexReverse = ( t1SwapIndex + 1 ) % 3;
		const otherEdge = triangle.edges[ otherEdgeIndex ].edge;
		const otherEdgeReverse = reverseTriangle.edges[ otherEdgeIndexReverse ].edge;
		const travlingEdge = { start: triangle.points[ t0SwapIndex ], end: reverseTriangle.points[ t1SwapIndex ] };
		const res = { x: NaN, y: NaN };
		const resReverse = { x: NaN, y: NaN };
		 closestPointLineToLine( otherEdge, travlingEdge, res );
		 if ( res.x <= 0 || res.x >= 1 ) {

			edge.required = true;

			return false;

		}

		 closestPointLineToLine( otherEdgeReverse, travlingEdge, resReverse );
		 if ( resReverse.x <= EPSILON || resReverse.x >= 1 - EPSILON ) {

			edge.required = true;

			return false;

		}
		
		```